File: rust/src/api/legacy/recorded_transaction.rs
use flutter_rust_bridge::frb;
use serde::{Deserialize, Serialize};

use crate::api::structs::amount::ApiAmount;
use crate::api::structs::recipient::ApiRecipient;

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub enum RecordedTransaction {
    Incoming(RecordedTransactionIncoming),
    Outgoing(RecordedTransactionOutgoing),
    UnknownOutgoing(RecordedTransactionUnknownOutgoing),
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct RecordedTransactionIncoming {
    pub txid: String,
    pub amount: ApiAmount,
    pub confirmation_height: Option<u32>,
    pub confirmation_blockhash: Option<String>,
}

impl RecordedTransactionIncoming {
    #[frb(sync)]
    pub fn to_string(&self) -> String {
        serde_json::to_string_pretty(&self).unwrap()
    }
}

impl RecordedTransactionOutgoing {
    #[frb(sync)]
    pub fn to_string(&self) -> String {
        serde_json::to_string_pretty(&self).unwrap()
    }

    #[frb(sync)]
    pub fn total_outgoing(&self) -> ApiAmount {
        let sum: u64 = self.recipients.iter().map(|r| r.amount.0).sum();
        // include fee to the total as well
        let fee = self.fee.0;

        ApiAmount(sum + fee)
    }
}

impl RecordedTransactionUnknownOutgoing {
    #[frb(sync)]
    pub fn to_string(&self) -> String {
        serde_json::to_string_pretty(&self).unwrap()
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct RecordedTransactionOutgoing {
    pub txid: String,
    pub spent_outpoints: Vec<crate::api::structs::outpoint::OutPoint>,
    pub recipients: Vec<ApiRecipient>,
    pub confirmation_height: Option<u32>,
    pub confirmation_blockhash: Option<String>,
    pub change: ApiAmount,
    pub fee: ApiAmount,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct RecordedTransactionUnknownOutgoing {
    pub amount: ApiAmount,
    pub confirmation_height: u32,
    pub confirmation_blockhash: Option<String>,
    pub spent_outpoints: Vec<crate::api::structs::outpoint::OutPoint>,
}

File: lib/data/models/recorded_transaction.dart
import 'package:danawallet/generated/rust/api/structs/amount.dart';
import 'package:danawallet/generated/rust/api/structs/outpoint.dart';
import 'package:danawallet/generated/rust/api/structs/recipient.dart';

sealed class RecordedTransaction {
  final int id;
  final String? note;
  final int? confirmationTimestamp;

  RecordedTransaction({
    required this.id,
    this.note,
    this.confirmationTimestamp,
  });

  String get txid;
  int? get confirmationHeight;
  String? get confirmationBlockhash;
}

final class RecordedTransactionIncoming extends RecordedTransaction {
  @override
  final String txid;
  final Amount amount;
  @override
  final int? confirmationHeight;
  @override
  final String? confirmationBlockhash;

  RecordedTransactionIncoming({
    required super.id,
    super.note,
    super.confirmationTimestamp,
    required this.txid,
    required this.amount,
    this.confirmationHeight,
    this.confirmationBlockhash,
  });
}

final class RecordedTransactionOutgoing extends RecordedTransaction {
  @override
  final String txid;
  final List<OutPoint> spentOutpoints;
  final List<Recipient> recipients;
  @override
  final int? confirmationHeight;
  @override
  final String? confirmationBlockhash;
  final Amount change;
  final Amount fee;

  RecordedTransactionOutgoing({
    required super.id,
    super.note,
    super.confirmationTimestamp,
    required this.txid,
    required this.spentOutpoints,
    required this.recipients,
    this.confirmationHeight,
    this.confirmationBlockhash,
    required this.change,
    required this.fee,
  });

  Amount totalOutgoing() {
    final sum = recipients.fold(
      BigInt.zero,
      (acc, r) => acc + r.amount.field0,
    );
    return Amount(field0: sum + fee.field0);
  }

  @override
  String toString() {
    final buffer = StringBuffer();
    buffer.writeln('txid: $txid');
    buffer.writeln('recipients:');
    for (final r in recipients) {
      buffer.writeln('  ${r.paymentCode}: ${r.amount.field0} sat');
    }
    buffer.writeln('fee: ${fee.field0} sat');
    buffer.writeln('change: ${change.field0} sat');
    buffer.write('confirmation_height: ${confirmationHeight ?? 'unconfirmed'}');
    return buffer.toString();
  }
}

final class RecordedTransactionUnknownOutgoing extends RecordedTransaction {
  final Amount amount;
  @override
  final int confirmationHeight;
  @override
  final String? confirmationBlockhash;
  final List<OutPoint> spentOutpoints;

  @override
  String get txid => 'Unknown';

  RecordedTransactionUnknownOutgoing({
    required super.id,
    super.note,
    super.confirmationTimestamp,
    required this.amount,
    required this.confirmationHeight,
    this.confirmationBlockhash,
    required this.spentOutpoints,
  });
}
